### ISSUE TITLE
Consideration of test leakage?

### ISSUE THREAD

**Issue by cthoyt** [OpenBioLink/OpenBioLink#9]
In section 4.3 of [Observed versus latent features for knowledge base and text inference](https://www.aclweb.org/anthology/W15-4007.pdf), Toutanova and Chen describe the issue of generating training/test sets of edges from networks that contain inverse relationships - if the triple `(h, r, t)` is in the training test set and its corresponding inverse triple `(t, r', h)` is in the test set, the learner will be able to easily pick up the pattern and there will likely be an overly-optimistic evaluation.

I think this is applicable both to random splitting and to some extent time-based splitting. Since this represents a significant advancement in the generation of benchmark datasets, I would encourage you check their workflow (which is described pretty well in the paper) and include an implementation in your package.

**Comment by matthias-samwald** [comment]
Yes, we took great care to not have this kind of trivially inferred relationships (inverse relations, super-relations) in the test set!

**Comment by cthoyt** [comment]
Awesome! How would this work for anyone who wants to take the same care with generation of new benchmark data sets using your package?

It would be worth mentioning that you did this in the manuscript and referencing this paper to serve as a proper warning for downstream users on the dangers that can come from this. For example, the machine learning community loves using the UMLS hierarchy as a benchmark dataset, but I think very few understand the kinds of relationships in it and they are often making unfair evaluations on it.

**Comment by matthias-samwald** [comment]
This is briefly mentioned (as much as the 2-page limit allowed) in the manuscript.

"Special focus was put on the robustness and difficulty of the test set, which
contains only entities that are also present in the training set and does not
contain relations that can be trivially inferred from the training set (such as
reverse edges of symmetric relations, inverse relations or super-relations)."

Filtering of these trivially inferred relations from the test set is done by the scripts that generate the train-test split. 

@annabreit Do you want to add something here?

**Comment by cthoyt** [comment]
I've also written an application note for Bioinformatics. I found it very tricky to get the right kind of information into the manuscript, and totally understand.

Since this is about a software application, these types of explanations (in long form) are crucial. I've found they're pretty effective when including them in full in the documentation, along with code examples on how to deal with the problems in practice. I also really enjoyed this talk when thinking about how to write documentation (that was more like writing a manuscript in the end) https://www.youtube.com/watch?v=t4vKPhjcMZg. Documentation is also a place where you can be much more liberal/granular in references, since these also take a lot of space in the manuscript and I found myself skipping a few when I did one...

I wouldn't say I'm perfect either, but you could consider the difference between my paper describing PyBEL (https://academic.oup.com/bioinformatics/article/34/4/703/4557184) and the documentation on ReadTheDocs (https://pybel.readthedocs.io) as an example.
